Refactor PSSS weekly reports to react query

tcaiger opened the pull request:
This is mostly a refactor to set up the data saving pattern for the app. It does also include issue [1584](https://github.com/beyondessential/tupaia-backlog/issues/1584) to add another editable field to the weekly-reports panel which was only a small change after the refactor.



### Changes:

- Moved the weekly-reports data fetching from redux to react-query
- Removed old redux and consolidated remaining weekly-reports redux code into a single file
- Added handling for saving data using react-query mutations 
- Added totalSites as an editable field

Note: there is another card for adding the logic for the current week [here](https://github.com/beyondessential/tupaia-backlog/issues/1628)

There is also another card for api integration so some of that might change.

---

edmofro commented on packages/psss/src/containers/Panels/WeeklyReportsPanel.js:
```diff
@@ -124,38 +130,54 @@ const toCommaList = values =>
+
 export const WeeklyReportsPanelComponent = React.memo(
-  ({ countryData, sitesData, isOpen, handleClose, unVerifiedSyndromes, alerts, handleConfirm }) => {
+  ({ countryData, isOpen, handleClose, weekNumber, unVerifiedSyndromes }) => {
```
What do you think about sending through just the country data for the week in question? I realise we're also passing in the weekNumber, so it's not totally necessary, but it's also more info than we need within this component, and means having to pass in an extra key every time we want to access the data within the component.

edmofro replied:
(this would be done in `mapStateToProps`)

edmofro commented on packages/psss/src/containers/Panels/WeeklyReportsPanel.js:
```diff
@@ -124,38 +130,54 @@ const toCommaList = values =>
         }
-
         setPanelStatus(PANEL_STATUSES.SUCCESS);
+        queryCache.invalidateQueries('country-weeks', { countryCode, weekNumber });
```
I'm wondering if there's a better way to manage what queries get invalidate by what mutations. E.g. a mutation to a single site's data would mutate the site query as well as the country query. Seems like the kind of thing we'd want to have centrally somehow, because down at the component level isn't really the place to be thinking of "what other parts of the app will be affected by this change"

edmofro reviewed (changes requested):
I know you're pretty keen on pushing stuff down into the components, but I'm still keen on keeping an extra layer of separation in there. 

The react query cache invalidation is so so so cool, I love it, solves a problem I've dealt with many times in different ways really nicely. I'm just wondering if we can have a central mapping between post endpoints and the get endpoints they invalidate? 

I think maybe both things could be solved together. I'm a bit distracted by being on a Tamanu call with Fiji at the moment, so don't have headspace to propose a solution, but would love if you can think that stuff over!

edmofro commented:
By the way I do agree this is a great improvement, and I am really enjoying learning some of this new react stuff through these PRs :-)

tcaiger commented on packages/psss/src/store/weeklyReports.js:
```diff
@@ -0,0 +1,70 @@
+
+// reducer
+const defaultState = {
+  id: null, // Todo: update to be the latest week by default??
```
ok thanks for that. I've made a note in the card for handling weeks to that extent.

tcaiger commented on packages/psss/src/containers/Tables/SiteReportTable.js:
```diff
@@ -54,57 +55,59 @@ const TABLE_STATUSES = {
-    await onSubmit(fields);
-    setTableStatus(TABLE_STATUSES.STATIC);
-  };
+  const [saveReport] = useMutation(
+    () => {
+      setTableStatus(TABLE_STATUSES.SAVING);
+      FakeAPI.post(fields);
+    },
+    {
+      onSuccess: () => queryCache.invalidateQueries('country-weeks', { countryCode, weekNumber }),
+      onSettled: () => setTableStatus(TABLE_STATUSES.STATIC),
+    },
+  );
```
I don't entirely agree. I think outside of the framework of redux, there isn't really the concept of "presentational" components anymore. I prefer to think of components as a piece of fully functional encapsulated and declarative ui. I don't really think that refetching data constitutes business logic and I think that other components using 'country-weeks' data are not impacted by refetching the data, they just display the latest available data from the server (which is the true centralised state). For point B, the api wont be a singleton any more after the api layer refactor.

tcaiger replied:
ps. It might be good to add the queryKey strings as constants.

tcaiger replied:
In saying that another option for abstracting some of this logic out side of the component, but without mapping the components together, would be to create a custom hook.

edmofro replied:
Heh, I'm probably going to take a bit of convincing to come along for the ride on this.

What I mean about the 'country-weeks' is that you'd have situations where a certain component makes a change that invalidates data it doesn't fetch, but other components do. For example, the component responsible for updating a site's weekly case count of AFR would need to invalidate the 'country-weeks' endpoint too, even though it otherwise has no idea about country level weekly data. Understanding that seems well outside of the realm of the specific component.

On point B - I don't think it makes a difference if it's an api singleton or functions (potentially wrapped in an object), it still means that you have to do more funky stubbing than just injecting a prop if you want to be able to use/demo/test the component in isolation right?

That's about enough blood shed on this hill for me! I'm happy to be talked into the modern age if you're still keen on keeping fetching and query invalidation inside these components. Would love a third opinion from @mcccclean, though I think he's probably too busy with Tamanu atm.

edmofro replied:
Your hook comment wasn't there when I wrote my reply - that seems a good modern pattern to pull out the bits I'm concerned about!

tcaiger commented on packages/psss/src/store/weeklyReports.js:
```diff
@@ -0,0 +1,70 @@
+
+    return list;
+  }, []);
+};
```
Yes I agree - good catch

tcaiger commented on packages/psss/src/views/Tabs/WeeklyCasesTabView.js:
```diff
@@ -66,90 +61,85 @@ const DateSubtitle = styled(Typography)`
+        />
+        {isFetching && 'Fetching...'}
+        {!isLoading && activeWeek !== null && (
+          <WeeklyReportsPanel activeWeek={activeWeek} countryWeekData={data.data[activeWeek]} />
```
Just pass in the the active week data that we need here

tcaiger commented:
@edmofro - This is ready for re-review. I have extracted the posting & re-fetching logic out into custom hooks in api/mutations.js file and also updated weekly reports to just use the week of data that it needs. Thanks for the great suggestions.

edmofro commented on packages/psss/src/containers/Panels/WeeklyReportsPanel.js:
```diff
@@ -281,20 +266,24 @@ export const WeeklyReportsPanelComponent = React.memo(
-  unVerifiedSyndromes: getUnVerifiedSyndromes(state, countryData),
-});
+const mapStateToProps = (state, { countryWeekData }) => {
+  const syndromeAlerts = countryWeekData.syndromes.filter(s => s.isAlert);
```
Great spot to put this!

edmofro reviewed (approved):
All looks really nice. I can see the slight tension of having a hook pulled out that wants to set state on the component which is also managed in other parts of the component, but you've handled it gracefully (and I quite like the return to try/catch around an await).

Awesome work, and thanks for pushing our use of react forward as always!
